pkg/controllers/tagging/tagging_controller.go

```diff
@@ -92,24 +95,44 @@ func NewTaggingController(
 	cloud cloudprovider.Interface,
 	nodeMonitorPeriod time.Duration,
 	tags map[string]string,
-	resources []string) (*Controller, error) {
+	resources []string,
+	rateLimit float64,
+	burstLimit int) (*Controller, error) {
 
 	awsCloud, ok := cloud.(*awsv1.Cloud)
 	if !ok {
 		err := fmt.Errorf("tagging controller does not support %v provider", cloud.ProviderName())
 		return nil, err
 	}
 
+	var rateLimiter workqueue.RateLimiter
+	var rateLimitEnabled bool
+	if rateLimit > 0.0 && burstLimit > 0 {
+		klog.Infof("Rate limit enabled on controller with rate %f and burst %d.", rateLimit, burstLimit)
+		// This is the workqueue.DefaultControllerRateLimiter() but in case where throttling is enabled on the controller,
+		// the rate and burst values are set to the provided values.
+		rateLimiter = workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(rateLimit), burstLimit)},
+		)
+		rateLimitEnabled = true
+	} else {
+		klog.Infof("Rate limit disabled on controller.")
+		rateLimiter = workqueue.DefaultControllerRateLimiter()
```

Comment:
Looks like this `DefaultControllerRateLimiter` defaults to the same pattern as what you have in the "if" block above, with the only difference being rate=10 and burst=100 ([xref](https://github.com/kubernetes/client-go/blob/v0.24.1/util/workqueue/default_rate_limiters.go#L43))

In that case, it might just be simpler to default the rate/burst flags to those values. Should achieve the same effect, without needing to if-else based on `rateLimitEnabled` in this PR?

Also - is 10 a good default for the rate given we know ec2 default is 5 qps?

Reply:
The reason for this if check is for backwards compatibility. We don't have any rate limiting today so we certainly didn't want to add rate limit by default.

As for the default values, they're only used for items running into errors when there's no rate limit enabled so the values seemed good because there's anyway a item-wise rate limit specified too. I am also not keeping these too coupled with EC2 values for two reasons:
1. This is a generic controller so the users of these controllers might have increased rate limits for EC2.
2. We'll support tagging more resources in the future so the rate limit of those services might be different than EC2.

I think these should be good values to begin with. We can tune these as and when required.

Reply:
Makes sense that you don't want to limit the qps by default.

Can we avoid forking this code artificially based on `rateLimitEnabled` though? It's being done at few different places in the PR. So IIUC you can achieve the above default by:
- always initialize workqueue as `NewMaxOfRateLimiter(NewItemExponentialFailureRateLimiter, BucketRateLimiter(rate, burst))`
- set the rate-limit flag as [float.Inf](https://pkg.go.dev/golang.org/x/time/rate#pkg-constants) by default (burst can be anything, even 0)
- always call `AddRateLimited` (even for new entries) everywhere in the controller code

Reply:
I had thought about this too. The primary issue with this is that while it works for the non-error case, for error cases, it deviates from the current behavior. 

Currently, whenever there's an error, an item is added back to the queue in a rate-limited way (using `NewMaxOfRateLimiter(NewItemExponentialFailureRateLimiter, BucketRateLimiter(rate, burst))`)

If we do the above, then in case of error cases, we'll effectively be disabling the `BucketRateLimiter` (because rate-limit is `float.Inf`) and hence items will only be rate-limited using the `ExponentialFailureRateLimiter` which is different from the above behavior.

While the difference isn't a lot (after a new secs the ExponentialRateLimiter will likely catch up), it is still backwards incompatible with whatever we have in our release today and hence my decision to go ahead with this approach.